### DISCUSSION TITLE
relative path in native file binaries section

### DISCUSSION THREAD

**Discussion by vbextreme** [mesonbuild/meson#11903] [Q&A]
I need use compiler in relative path inside native file, examples, ./tools/sdk/compile/gcc, exists one methods?
for now I use full path but is not good

now
```
[constants]
arch = 'esp32'
repo = '/home/username/project/embedded/c'
path = repo / 'tools/esp/sdk/' + arch + '/xtensa-' + arch + '-elf/bin'

[binaries]
c = [ path / 'xtensa-' + arch + '-elf-gcc' ]
```

help me to auto set repo in [constants]

**Comment by xclaesse** [comment]
This is discussed in https://github.com/mesonbuild/meson/pull/9223

**Comment by vbextreme** [comment]
I have already readed.
do you think we could remove those pesky bootstrap scripts in the future?
A good idea Is to have default constans in native file, for examples CURRENT_PATH.

bin=CURRENT_PATH / CURRENT_HOST / 'sdk' / 'bin'
...
gcc=bin / 'mcu-gcc'

Thanks for your time.
Have a good life.

**Reply by xclaesse** [reply]
I hesitated adding `machine_file_path` (name TBD) in https://github.com/mesonbuild/meson/pull/12083 too. The issue is we can have multiple machines files in different locations, and the way it's currently implemented values are evaluated after all files have been merged. That means we would have to assume `machine_file_path` is the path of the first in the list... Not great but probably what people wants I'm most cases.

**Comment by vbextreme** [comment]
The problem must be analyzed first:
When do we need the relative path of the native file?
We need it solely and exclusively when we want to use the compiler or other software in a directory relative to the native file itself.
Resolving all other paths can be done comfortably within the meson.build; therefore, doing it in the native file would be unnecessary.

The problem can be solved in two ways.
The first and probably simpler way would be to take the path of the native file where machine_file_path is encountered for the first time. There might be multiple native files, but only one of them will have [binaries].
The second, probably less elegant, way would be to provide a method inside the native file to indicate when to set the machine_file_path, but honestly, I don't like that idea.

By using the first approach, we can do something interesting like:
```
$ meson --cross-file ~/cross/avr-328p-16mhz.ini --cross-file toolkit/avr/bin/avr-gcc-5.ini
```
where the order of the calls wouldn't affect the result, so it would suffice to do:
```
$ meson --cross-file ~/cross/avr-328p-16mhz.ini
```
to compile it with the compiler installed on the OS.

It could also be possible to restrict the usage of machine_file_path only within the [binaries] section.


Thank you for the time you've dedicated to me.

**Reply by xclaesse** [reply]
There is a lengthy discussion about this starting from https://github.com/mesonbuild/meson/pull/12083#issuecomment-1671177540.

> only one of them will have [binaries]

That's not guaranteed

> take the path of the native file where machine_file_path is encountered for the first time

It is mostly an implementation detail, how it works with python's configparser, we don't know which files uses `machine_file_path`.

**Reply by vbextreme** [reply]
You know the Paths in advance, you can create a variable for each `--cross-file` flag, with the same name as the `file.ini`. This way, `cross/any-bar.ini` will become: `native_path_cross_any_bar`. This will not only allow you to safely access the actual directory, but also use the Path of another native file, eliminating the need to have an order for native files.
More better
```
foo = native_path['cross/any-bar.ini']
foo = native_path['@1@']
```
Where @N@ is order --cross-file
